src/handle_init_contract.c

```diff
@@ -0,0 +1,94 @@
+/*******************************************************************************
+ *
+ * ██╗  ██╗██╗██╗     ███╗   ██╗
+ * ██║ ██╔╝██║██║     ████╗  ██║
+ * █████╔╝ ██║██║     ██╔██╗ ██║
+ * ██╔═██╗ ██║██║     ██║╚██╗██║
+ * ██║  ██╗██║███████╗██║ ╚████║
+ * ╚═╝  ╚═╝╚═╝╚══════╝╚═╝  ╚═══╝
+ *
+ * Kiln Ethereum Ledger App
+ * (c) 2022-2024 Kiln
+ *
+ * contact@kiln.fi
+ ********************************************************************************/
+
+#include "kiln_plugin.h"
+#include "plugin_utils.h"
+
+void handle_init_contract(ethPluginInitContract_t *msg) {
+    if (msg->interfaceVersion != ETH_PLUGIN_INTERFACE_VERSION_LATEST) {
+        msg->result = ETH_PLUGIN_RESULT_UNAVAILABLE;
+        return;
+    }
+
+    if (msg->pluginContextLength < sizeof(context_t)) {
+        PRINTF("Plugin parameters structure is bigger than allowed size\n");
+        msg->result = ETH_PLUGIN_RESULT_ERROR;
+        return;
+    }
+
+    context_t *context = (context_t *) msg->pluginContext;
+
+    memset(context, 0, sizeof(*context));
+
+    size_t index;
+    if (!find_selector(U4BE(msg->selector, 0), KILN_SELECTORS, NUM_SELECTORS, &index)) {
+        PRINTF("Error: selector not found!\n");
+        msg->result = ETH_PLUGIN_RESULT_UNAVAILABLE;
+        return;
+    }
+    context->selectorIndex = index;
+    // check for overflow
+    if ((size_t) context->selectorIndex != index) {
```

Comment:
This looks a bit weird hard to grasp as it plays with overflows to detect them, integer overflow is actually undefined behaviour so it's a bit weird to have logic/assumptions around it. I think it'd make more sense to do:

```
if (context->selectorIndex >= NUM_SELECTORS} {
        PRINTF("Error: overflow detected on selector index!\n");
        msg->result = ETH_PLUGIN_RESULT_ERROR;
      return;
}
```

It will also catch more error as if say, find_selector returns 234, it won't overflow but still crash later as we'll out of bounds.

Reply:
Actually the switch case below catches it as well. I'm still unsure about this check's intent, I don't have full context and maybe there is a reason, without context I'd remove it as it's confusing/weird, and still covered by the switch case below.